### PR TITLE
[WIP-DNM] ectd (ticdc): add etcd meta data magration logic

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #xxx

#### Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

    test case 1: 
      1. run a cdc cluster with 3 old cdc server
      2. update **owner node**
      3. update **processor node**
      4. migrate successful 

    test case 2: 
      1. run a cdc cluster with 3 old cdc server
      2. update **pocessor nodes**
      3. update **owner node**
      4. migrate successful 

    test case 2: 
      1. run a cdc cluster with 3 old cdc server
      2. tiup cluster patch  and restart all cdc nodes
      3. migrate successful 

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

File: pkg/orchestrator/etcd_worker.go [hunk 1]
```diff
@@ -121,9 +121,20 @@ func (worker *EtcdWorker) initMetrics() {
 // And the specified etcd session is nil-safety.
 func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session, timerInterval time.Duration, role string) error {
 	defer worker.cleanUp()
+
+	if role == "owner" {
```
Comment (sdojjy) [inline pkg/orchestrator/etcd_worker.go:125]: use `util.RoleOwner.String()`

File: pkg/etcd/migrate.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	}
+	election := concurrency.NewElection(sess, m.scrOwnerKey)
+	electionCtx, cancel := context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
+	defer func() {
+		_ = sess.Close()
+	}()
+
+	// TODO(dongmen): do more investigration about how to handle error here
+	if err := election.Campaign(electionCtx, migrationCampaignKey); err != nil {
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:119]: should be compaign the cdc owner key?

File: pkg/etcd/migrate.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	m := migration{
+		oldMetaVersion:   oldVersion,
+		newMetaVersion:   cdcMetaVersion,
+		metaVersionKey:   DefaultClusterAndMetaPrefix + metaVersionKey,
+		scrOwnerKey:      "/ticdc/cdc/owner",
+		desCli:           cli,
+		keys:             make(keys),
+		srcNoMetaVersion: noMetaVersionKey,
+	}
+	m.keys.addPair("/tidb/cdc/changefeed/info", DefaultClusterAndNamespacePrefix+changefeedInfoKey)
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:166]: remove this line,  prefix is used in `migrate` func

File: pkg/etcd/migrate.go (hunk 2 above)
Comment (asddongmen) [reply pkg/etcd/migrate.go:119]: It shouldn't, etcd will campaign on election's prefix key we set as `m.oldOwnerKey` above:
```
election := concurrency.NewElection(sess, m.oldOwnerKey)
```

File: pkg/orchestrator/etcd_worker.go (hunk 1 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:125]: ok

File: pkg/orchestrator/etcd_worker.go [hunk 4]
```diff
@@ -512,3 +523,25 @@ func (worker *EtcdWorker) handleDeleteCounter(value []byte) {
...
+	}
+	if !shouldMigrate {
+		return nil
+	}
+
+	if !worker.isOwner {
+		err := etcd.WaitMetaVersionMatched(ctx, worker.client)
+		if err != nil {
+			return errors.Trace(err)
+		}
```
Comment (sdojjy) [inline pkg/orchestrator/etcd_worker.go:542]: add return here, otherwise , processor will do migration too

File: pkg/etcd/migrate.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			}
+			if version == cdcMetaVersion {
+				return nil
+			}
+		}
+	}
+}
+
+func getMetaVersion(ctx context.Context, cli *Client) (int, error) {
+	key := DefaultClusterAndMetaPrefix + metaVersionKey
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:220]: we can get the cluster id from cmd parameter, so we should not use `DefaultClusterAndMetaPrefix`, although the cluster id is "default"

File: pkg/etcd/migrate.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+type migrater struct {
+	oldMetaVersion int
+	newMetaVersion int
+	metaVersionKey string
+	// cdc old onwer key
+	oldOwnerKey string
+	// etcd client
+	cli *Client
+	// all keyPrefixs needed to be migrated or update
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:52]:
```suggestion
	// all keyPrefixes needed to be migrated or update
```

File: pkg/etcd/migrate.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	// map from oldKeyPrefix to newKeyPrefix
+	keyPrefixs        keys
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:54]:
```suggestion
	keyPrefixes        keys
```

File: pkg/etcd/migrate.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+	migrationCampaignKey = "ticdc-migration"
+)
+
+type keys map[string]string
+
+func (k keys) addPair(old, new string) {
+	k[old] = new
+}
+
+type migrater struct {
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:44]:
```suggestion
type migrator struct {
```

File: pkg/etcd/migrate.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+// 4. update metaVersion
+func (m migrater) migrate(ctx context.Context) error {
+	// 1.1 check metaVersion, if the metaVersion in etcd does not match
+	// m.oldMetaVersion, it means that someone has migrated the meta data
+	metaVersion, err := getMetaVersion(ctx, m.cli)
+	if err != nil {
+		log.Error("get meta version failed, etcd meta data migration failed", zap.Error(err))
+		return errors.Trace(err)
+	}
+	if !m.etcdNoMetaVersion && metaVersion != m.oldMetaVersion {
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:73]: compare  metaVersion  with m.newMetaVersion ?
`if metaVersion == m.newMetaVersion `

File: pkg/etcd/migrate.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+			// TODO(dongmen): we should update value after changedfeedInfo struct altered
+			_, err := m.cli.Put(ctx, newKey, string(v.Value))
+			if err != nil {
+				log.Error("put new meta data failed, etcd meta data migration failed", zap.Error(err))
+				return errors.Trace(err)
+			}
+		}
+	}
+
+	// 5. update meataVersion
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:112]:
```suggestion
	// 5. update metaVersion
```

File: pkg/etcd/migrate.go [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+func (k keys) addPair(old, new string) {
+	k[old] = new
+}
+
+type migrater struct {
+	oldMetaVersion int
+	newMetaVersion int
+	metaVersionKey string
+	// cdc old onwer key
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:48]:
```suggestion
	// cdc old owner key
```

File: pkg/etcd/migrate.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+func ShouldMigrate(ctx context.Context, cli *Client) (bool, error) {
+	version, err := getMetaVersion(ctx, cli)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return version != cdcMetaVersion, nil
+}
+
+// WaitMetaVersionMatched checks and waits until the metaVersion in etcd
+// matched to loack cdcMetaVersion
```
Comment (sdojjy) [inline pkg/etcd/migrate.go:192]: loack,  typo?

File: pkg/orchestrator/etcd_worker.go [hunk 13]
```diff
@@ -120,9 +120,20 @@ func (worker *EtcdWorker) initMetrics() {
 // And the specified etcd session is nil-safety.
 func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session, timerInterval time.Duration, role string) error {
 	defer worker.cleanUp()
+
+	if role == pkgutil.RoleOwner.String() {
```
Comment (sdojjy) [inline pkg/orchestrator/etcd_worker.go:124]: maybe we can put these 3 line into checkAndMigrateMetaData func

File: pkg/orchestrator/etcd_worker.go (hunk 4 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:542]: good catch

File: pkg/orchestrator/etcd_worker.go (hunk 13 above)
Comment (asddongmen) [reply pkg/orchestrator/etcd_worker.go:124]: ok

File: pkg/etcd/migrate.go [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+	// 1.1 check metaVersion, if the metaVersion in etcd does not match
+	// m.oldMetaVersion, it means that someone has migrated the meta data
+	metaVersion, err := getMetaVersion(ctx, m.cli, m.clusterID)
+	if err != nil {
+		log.Error("get meta version failed, etcd meta data migration failed", zap.Error(err))
+		return errors.Trace(err)
+	}
+
+	if metaVersion > m.newMetaVersion {
+		log.Panic("meta version in etcd is greater than the meta version in TiCDC",
```
Comment (liuzix) [inline pkg/etcd/migrate.go:76]: Will this be triggered under normal circumstances?

File: pkg/etcd/migrate.go [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+			zap.Int("etcdMetaVersion", metaVersion), zap.Int("cdcMetaVersion", m.newMetaVersion))
+	}
+
+	// if metaVersion in etcd is equal to m.newMetaVersion,
+	// it means that no need to migrate
```
Comment (liuzix) [inline pkg/etcd/migrate.go:81]:
```suggestion
	// it means that there is no need to migrate
```

File: pkg/etcd/migrate.go [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+	if m.etcdNoMetaVersion {
+		_, err := m.cli.Put(ctx, m.metaVersionKey, fmt.Sprintf("%d", m.oldMetaVersion))
+		if err != nil {
+			log.Error("put meta version failed, etcd meta data migration failed", zap.Error(err))
+			return errors.Trace(err)
+		}
+	}
+
+	// 3.campaign old owner to make sure old keys will not be updates
+	if err := m.campaignOldOwner(ctx); err != nil {
```
Comment (liuzix) [inline pkg/etcd/migrate.go:97]: Do we need a timeout here? If the current node cannot become the old owner, should it print some useful logs?

File: pkg/etcd/migrate.go [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+		if err != nil {
+			log.Error("get old meta data failed, etcd meta data migration failed", zap.Error(err))
+			return errors.Trace(err)
+		}
+		for _, v := range resp.Kvs {
+			oldKey := string(v.Key)
+			newKey := newPrefix + oldKey[len(oldPrefix):]
+			log.Info("migrate key", zap.String("oldKey", oldKey), zap.String("newKey", newKey))
+			// TODO(dongmen): we should update value after changefeedInfo struct altered
+			_, err := m.cli.Put(ctx, newKey, string(v.Value))
```
Comment (liuzix) [inline pkg/etcd/migrate.go:114]: Should we check lease atomically with the `Put` operation? Can the current node's lease suddenly expire doing these operations? If so, is data corruption possible?

File: pkg/orchestrator/etcd_worker.go [hunk 18]
```diff
@@ -508,3 +520,24 @@ func (worker *EtcdWorker) handleDeleteCounter(value []byte) {
...
+	shouldMigrate, err := etcd.ShouldMigrate(ctx, worker.client, clusterID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !shouldMigrate {
+		return nil
+	}
+
+	if role != pkgutil.RoleOwner.String() {
+		log.Info("gaga", zap.String("role", role))
```
Comment (liuzix) [inline pkg/orchestrator/etcd_worker.go:536]:
```suggestion
```

File: pkg/etcd/migrate.go (hunk 14 above)
Comment (asddongmen) [reply pkg/etcd/migrate.go:76]: It will never happen in a normal circumstance.

File: pkg/etcd/migrate.go (hunk 16 above)
Comment (asddongmen) [reply pkg/etcd/migrate.go:97]: ok

File: pkg/etcd/migrate.go (hunk 17 above)
Comment (asddongmen) [reply pkg/etcd/migrate.go:128]: Normally, migrate will exit before cdc node's lease expired and only a cdc node with double ownership can put new meta data to etcd, so I consider there is no risk of data corruption.

**Pull request merged**